File: src/shared/roadtrip/hooks/picked.ts
/**
 * The pictures an author picked for an opener, as they are stored in its
 * options — read defensively, and ordered the way they were shot.
 *
 * Grown in Défilé, lifted out when the drive wanted the same list with one
 * more thing on it: the POSITION a picture was shot at, so a geotagged
 * picture can be a stop on a map. Défilé re-exports the old names and
 * ignores the position. Pure and DOM-free.
 */

import { hookPictureKey, type HookDay, type HookPickedPicture } from './hook-variant';

const ISO_DAY = /^\d{4}-\d{2}-\d{2}$/;

/** A stored position, or null: two finite numbers inside the globe's bounds. */
export function readCoords(raw: unknown): { lat: number; lon: number } | null {
  if (!raw || typeof raw !== 'object') return null;
  const { lat, lon } = raw as Record<string, unknown>;
  if (typeof lat !== 'number' || typeof lon !== 'number') return null;
  if (!Number.isFinite(lat) || !Number.isFinite(lon)) return null;
  if (Math.abs(lat) > 90 || Math.abs(lon) > 180) return null;
  return { lat, lon };
}

/**
 * A stored picked list, read defensively: an entry with no readable ref or
 * day is dropped, a second entry for the same picture is dropped, a position
 * that is not two finite numbers is dropped. What a newer build wrote beside
 * the known keys is left behind, not trusted.
 */
export function readPicked(raw: unknown): HookPickedPicture[] {
  if (!Array.isArray(raw)) return [];
  const seen = new Set<string>();
  const out: HookPickedPicture[] = [];
  for (const item of raw) {
    if (!item || typeof item !== 'object') continue;
    const { ref, date, takenAt, coords } = item as Record<string, unknown>;
    if (typeof date !== 'string' || !ISO_DAY.test(date)) continue;
    if (!ref || typeof ref !== 'object') continue;
    const r = ref as Record<string, unknown>;
    if (typeof r.name !== 'string' || !r.name) continue;
    const clean = {
      name: r.name,
      size: Number.isFinite(r.size) ? Number(r.size) : 0,
      lastModified: Number.isFinite(r.lastModified) ? Number(r.lastModified) : 0,
      ...(typeof r.assetId === 'string' && r.assetId ? { assetId: r.assetId } : {}),
      ...(typeof r.hash === 'string' && r.hash ? { hash: r.hash } : {}),
    };
    const key = hookPictureKey(clean);
    if (seen.has(key)) continue;
    seen.add(key);
    const position = readCoords(coords);
    out.push({
      ref: clean,
      date,
      ...(typeof takenAt === 'number' && Number.isFinite(takenAt) ? { takenAt } : {}),
      ...(position ? { coords: position } : {}),
    });
  }
  return out;
}

/** Picked pictures in the order they were shot: day, then instant, then name. */
export function sortPicked(picked: readonly HookPickedPicture[]): HookPickedPicture[] {
  return [...picked].sort(
    (a, b) =>
      a.date.localeCompare(b.date) ||
      (a.takenAt ?? Number.MAX_SAFE_INTEGER) - (b.takenAt ?? Number.MAX_SAFE_INTEGER) ||
      a.ref.name.localeCompare(b.ref.name),
  );
}

/**
 * Where each picked picture falls against this piece: in reach (shot on a
 * day of the trip, no later than the piece's own), after it, or outside the
 * trip altogether. A panel says the last two out loud; a plan uses the first.
 */
export function partitionPicked(
  calendar: readonly HookDay[],
  date: string,
  picked: readonly HookPickedPicture[],
): { inReach: HookPickedPicture[]; after: number; outside: number } {
  const days = new Set(calendar.map((day) => day.date));
  const inReach: HookPickedPicture[] = [];
  let after = 0;
  let outside = 0;
  for (const picture of picked) {
    if (!days.has(picture.date)) outside += 1;
    else if (picture.date > date) after += 1;
    else inReach.push(picture);
  }
  return { inReach: sortPicked(inReach), after, outside };
}

/**
 * `k` items spread evenly over `items`, first and last always kept. Fewer than
 * `k` comes back whole: a sweep never repeats a day to reach a count.
 */
export function sampleEvenly<T>(items: readonly T[], k: number): T[] {
  if (k <= 0) return [];
  if (items.length <= k) return [...items];
  if (k === 1) return [items[items.length - 1]];
  const out: T[] = [];
  for (let i = 0; i < k; i++) {
    out.push(items[Math.round((i * (items.length - 1)) / (k - 1))]);
  }
  return out;
}

File: apple/Packages/AtelierKit/Sources/AtelierKit/Roadtrip/Hooks/Picked.swift
// The pictures an author picked for an opener, as they are stored in its
// options — read defensively, and ordered the way they were shot. Port of
// `src/shared/roadtrip/hooks/picked.ts`.
//
// Grown in Défilé, lifted out when the drive wanted the same list with one
// more thing on it: the POSITION a picture was shot at, so a geotagged
// picture can be a stop on a map. Défilé ignores the position.
//
// Rules kept:
// - A stored entry is never trusted: one with no readable ref or day is
//   dropped, a second entry for the same picture (by `hookPictureKey`) is
//   dropped, a position that is not two finite numbers inside the globe is
//   dropped. What a newer build wrote beside the known keys is left behind.
// - A ref's size and date are read only from JSON NUMBERS (the web's
//   `Number.isFinite`, which never coerces a string).

import Foundation

/// A stored position, or nil: two finite numbers inside the globe's bounds.
public func readCoords(_ raw: JSONValue?) -> GeoPoint? {
    guard let o = raw?.objectValue,
          case .number(let lat)? = o["lat"], case .number(let lon)? = o["lon"],
          lat.isFinite, lon.isFinite, abs(lat) <= 90, abs(lon) <= 180 else { return nil }
    return GeoPoint(lat: lat, lon: lon)
}

/// `/^\d{4}-\d{2}-\d{2}$/`.
private func isIsoDay(_ s: String) -> Bool {
    let b = Array(s.utf8)
    guard b.count == 10 else { return false }
    for (i, c) in b.enumerated() {
        if i == 4 || i == 7 {
            if c != UInt8(ascii: "-") { return false }
        } else if c < UInt8(ascii: "0") || c > UInt8(ascii: "9") {
            return false
        }
    }
    return true
}

/// A stored ref kept only when it names a file; its size and date only when
/// they are finite numbers.
private func pickedRef(_ raw: JSONValue?) -> SavedMediaRef? {
    guard let r = raw?.objectValue, let name = r["name"]?.stringValue, !name.isEmpty else { return nil }
    let size = r["size"]?.finiteNumber ?? 0
    var ref = SavedMediaRef(name: name, size: abs(size) < 9e15 ? Int(size) : 0,
                            lastModified: r["lastModified"]?.finiteNumber ?? 0)
    if let assetId = r["assetId"]?.stringValue, !assetId.isEmpty { ref.assetId = assetId }
    if let hash = r["hash"]?.stringValue, !hash.isEmpty { ref.hash = hash }
    return ref
}

/// A stored picked list, read defensively (see the header).
public func readPicked(_ raw: JSONValue?) -> [HookPickedPicture] {
    guard let list = raw?.arrayValue else { return [] }
    var seen = Set<String>()
    var out: [HookPickedPicture] = []
    for item in list {
        guard let o = item.objectValue else { continue }
        guard let date = o["date"]?.stringValue, isIsoDay(date) else { continue }
        guard let ref = pickedRef(o["ref"]) else { continue }
        let key = hookPictureKey(ref)
        if seen.contains(key) { continue }
        seen.insert(key)
        var takenAt: Double? = nil
        if case .number(let n)? = o["takenAt"], n.isFinite { takenAt = n }
        out.append(HookPickedPicture(ref: ref, date: date, takenAt: takenAt, coords: readCoords(o["coords"])))
    }
    return out
}

/// Picked pictures in the order they were shot: day, then instant (an unknown
/// one last), then name — a stable sort, as the web's.
public func sortPicked(_ picked: [HookPickedPicture]) -> [HookPickedPicture] {
    let unknown = 9_007_199_254_740_991.0 // Number.MAX_SAFE_INTEGER
    func order(_ a: HookPickedPicture, _ b: HookPickedPicture) -> Int {
        if a.date != b.date { return a.date < b.date ? -1 : 1 }
        let ta = a.takenAt ?? unknown
        let tb = b.takenAt ?? unknown
        if ta != tb { return ta < tb ? -1 : 1 }
        return packLocaleCompare(a.ref.name, b.ref.name)
    }
    return picked.enumerated()
        .sorted { x, y in
            let c = order(x.element, y.element)
            return c != 0 ? c < 0 : x.offset < y.offset
        }
        .map(\.element)
}

/// Where each picked picture falls against a piece.
public struct PickedPartition: Equatable, Sendable {
    /// Shot on a day of the trip, no later than the piece's own — in shot order.
    public var inReach: [HookPickedPicture]
    /// How many were shot after the piece's day.
    public var after: Int
    /// How many were shot outside the trip altogether.
    public var outside: Int

    public init(inReach: [HookPickedPicture], after: Int, outside: Int) {
        self.inReach = inReach; self.after = after; self.outside = outside
    }
}

/// Where each picked picture falls against this piece: in reach (shot on a
/// day of the trip, no later than the piece's own), after it, or outside the
/// trip altogether. A panel says the last two out loud; a plan uses the first.
public func partitionPicked(_ calendar: [HookDay], _ date: IsoDate, _ picked: [HookPickedPicture]) -> PickedPartition {
    let days = Set(calendar.map(\.date))
    var inReach: [HookPickedPicture] = []
    var after = 0
    var outside = 0
    for picture in picked {
        if !days.contains(picture.date) {
            outside += 1
        } else if picture.date > date {
            after += 1
        } else {
            inReach.append(picture)
        }
    }
    return PickedPartition(inReach: sortPicked(inReach), after: after, outside: outside)
}

/// `k` items spread evenly over `items`, first and last always kept. Fewer
/// than `k` comes back whole: a sweep never repeats a day to reach a count.
public func sampleEvenly<T>(_ items: [T], _ k: Int) -> [T] {
    if k <= 0 { return [] }
    if items.count <= k { return items }
    if k == 1 { return [items[items.count - 1]] }
    var out: [T] = []
    out.reserveCapacity(k)
    for i in 0..<k {
        let index = TripJS.round(Double(i * (items.count - 1)) / Double(k - 1))
        out.append(items[Int(index)])
    }
    return out
}
